include/stdexec/execution.hpp

```diff
@@ -243,8 +294,35 @@ namespace stdexec {
           noexcept(nothrow_tag_invocable<get_env_t, const _EnvProvider&>)
           -> tag_invoke_result_t<get_env_t, const _EnvProvider&> {
           static_assert(queryable<tag_invoke_result_t<get_env_t, const _EnvProvider&> >);
+          __check_sender_version<remove_cvref_t<_EnvProvider>>();
           return tag_invoke(*this, __with_env);
         }
+
+      // NOT TO SPEC: This overload is provided to support
+      // R5 sender types which don't yet define get_env or
+      // evaluate true for the enable_sender trait. Remove
+      // this overload when deprecating R5 support.
+      template <class _EnvProvider>
+          requires
+            (!tag_invocable<get_env_t, const _EnvProvider&>) &&
+            __r5_sender<_EnvProvider> &&
+            (!__r7_sender<_EnvProvider>)
+        constexpr auto operator()(const _EnvProvider& __with_env) const
+          noexcept -> const _EnvProvider& {
+          __check_sender_version<remove_cvref_t<_EnvProvider>>();
+          return __with_env;
+        }
+
+      template <class _EnvProvider>
+          requires
+            (!tag_invocable<get_env_t, const _EnvProvider&>) &&
+            // NOT TO SPEC: Remove the R5/R7 sender checks when
+            // deprecating R5 support.
+            (!__r5_sender<_EnvProvider>)
+        constexpr auto operator()(const _EnvProvider& __with_env) const
+          noexcept -> __empty_env {
+          return {};
+        }
```

Comment:
I _think_ these two overloads can be combined into one that basically does:

```c++
if constexpr (__r5_sender<_EnvProvider>)
  return __with_env;
else
  return __empty_env{};
```
... provided you make my suggested change to the `__r5_sender` concept above.

Reply:
This was interesting - if I use trailing return type of `__if_c<__r5_sender<_EnvProvider>, const _EnvProvider&, __empty_env>`, clang gave

```
error: satisfaction of constraint 'stdexec::__valid<exec::__task::awaiter_context_t, __promise>' depends on itself
```

GCC 12 permitted the conditional return type. I've left the return type as `decltype(auto)` for now.

Reply:
These sorts of errors can be tricky to track down, but it's usually worth it because it generally indicates that something is fishy somewhere. I'll try to have a look.

Reply:
I can have another go at this one and let you know if I'm hopeless stuck. I had a good lesson over the past couple days on the meaning of immediate vs non-immediate context regarding SFINAE.

Reply:
Maybe rebase on `main` and try again? I've fixed some issues that likely were causing more to be instantiated than necessary.

Reply:
Same result.

Reply:
I think what I have is problematic. When https://github.com/NVIDIA/stdexec/blob/main/include/exec/task.hpp#L313 evaluates the constraint, it ends up checking `env_of_t`, which (with my changes) checks if `get_env`'s _EnvProvider is a sender, which checks if the type is awaitable, and we're back at line 313 evaluating the same constraint.

`get_env` is about as low level as can be, so checking if `_EnvProvider` is a sender risks cycles like this. Suggestions on how to proceed? My original approach of having this be two separate overloads "worked," but I'm not sure if that's even guaranteed given the apparent cycle.